packages/@aws-cdk/aws-batch/lib/job-definition.ts

```diff
@@ -0,0 +1,345 @@
+     * The name of the batch job definition
+     *
+     * @attribute
+     * @default Cloudformation-generated name
+     */
+    readonly jobDefinitionName: string;
+}
+
+/**
+ * Batch Job Definition
+ *
+ * Defines a batch job definition to execute a specific batch job.
+ */
+export class JobDefinition extends Resource implements IJobDefinition {
+    /**
+     * Fetches an existing batch job definition by its amazon resource name.
+     *
+     * @param scope
+     * @param id
+     * @param jobDefinitionArn
+     */
+    public static fromJobDefinitionArn(scope: Construct, id: string, jobDefinitionArn: string): IJobDefinition {
+        const stack = Stack.of(scope);
+        const jobDefName = stack.parseArn(jobDefinitionArn).resource;
+
+        class Import extends Resource implements IJobDefinition {
+            public readonly jobDefinitionArn = jobDefinitionArn;
+            public readonly jobDefinitionName = jobDefName;
+        }
+
+        return new Import(scope, id);
+    }
+
+    public readonly jobDefinitionArn: string;
+    public readonly jobDefinitionName: string;
+
+    constructor(scope: Construct, id: string, props: JobDefinitionProps = {}) {
+        super(scope, id, {
+            physicalName: props.jobDefinitionName,
+        });
+
+        const jobDef = new CfnJobDefinition(this, 'Resource', {
+            jobDefinitionName: props.jobDefinitionName,
+            containerProperties: this.buildJobContainer(props.containerProps),
+            type: 'container',
+            nodeProperties: props.nodeProps ? {
+                mainNode: props.nodeProps.mainNode,
+                nodeRangeProperties: this.buildNodeRangeProps(props.nodeProps),
+                numNodes: props.nodeProps.count,
+            } : undefined,
+            parameters: props.parameters,
+            retryStrategy: {
+                attempts: props.retryAttempts || 1,
+            },
+            timeout: {
+                attemptDurationSeconds: props.timeout ? props.timeout.toSeconds() : undefined,
+            },
+        });
+
+        this.jobDefinitionArn = jobDef.ref;
+        this.jobDefinitionName = this.getResourceNameAttribute(this.physicalName);
+    }
+
+    private deserializeEnvVariables(env?: { [name: string]: string}): CfnJobDefinition.EnvironmentProperty[] | undefined {
+        const vars = new Array<CfnJobDefinition.EnvironmentProperty>();
+
+        if (env === undefined) {
+            return vars;
+        }
+
+        Object.keys(env).map((name: string) => {
+            vars.push({ name, value: env[name] });
+        });
+
+        return vars;
+    }
+
+    private buildJobContainer(containerProps?: IJobDefinitionContainer): CfnJobDefinition.ContainerPropertiesProperty | undefined {
+        if (containerProps === undefined) {
+            return undefined;
+        }
+
+        return {
+            command: containerProps.command,
+            environment: this.deserializeEnvVariables(containerProps.environment),
```

Comment:
This is an optional property, but you always return an empty array, even if `props.environment` is `undefined`. That's not great - we generally don't fill properties that were not provided.

Reply:
Ok, so then would leaving it undefined be preferred?

Reply:
Yep.